File: backend/app/models/simulation.py
"""Pydantic models for the season-simulation debug API."""

from __future__ import annotations

from typing import Optional

from pydantic import BaseModel


class StatCell(BaseModel):
    value: float
    low: Optional[float] = None
    high: Optional[float] = None


class Game(BaseModel):
    game_id: str
    team_id: int
    team_abbr: str
    opponent_team_id: int
    opponent_abbr: str
    matchup: str


class SimState(BaseModel):
    season: str
    current_date: Optional[str]
    next_game_day: Optional[str]
    day_index: int
    total_days: int
    finished: bool
    num_games: int
    games: list[Game]


class PlayerPrediction(BaseModel):
    player_id: int
    player_name: str
    team_id: int
    team_abbr: str
    opponent_team_id: int
    opponent_abbr: str
    is_home: bool
    minutes: float
    default_minutes: float
    eligible: bool
    reason: str = ""
    stats: dict[str, StatCell] = {}


class EvalRow(BaseModel):
    player_id: int
    player_name: str
    team_abbr: str
    opponent_abbr: str
    is_home: bool
    real_minutes: float
    eligible: bool
    reason: str = ""
    predicted: dict[str, float] = {}
    actual: dict[str, float] = {}


class LastResults(BaseModel):
    played_date: Optional[str]
    evaluations: list[EvalRow]


class UpcomingResponse(BaseModel):
    state: SimState
    predictions: list[PlayerPrediction]
    last_results: Optional[LastResults] = None


class AdvanceResponse(BaseModel):
    played_date: Optional[str]
    evaluations: list[EvalRow]
    state: SimState


class PlayerStoreSummary(BaseModel):
    player_id: int
    player_name: str
    team_abbr: str
    games_count: int
    eligible: bool


class PlayersListResponse(BaseModel):
    current_date: Optional[str]
    next_game_day: Optional[str]
    players: list[PlayerStoreSummary]


class PlayerStoreState(BaseModel):
    player_id: int
    player_name: str
    team_abbr: str
    position: str
    last_game_date: Optional[str]
    games_count: int
    eligible: bool
    features: dict[str, Optional[float]]


class TeamSummary(BaseModel):
    team_id: int
    team_abbr: str


class TeamsListResponse(BaseModel):
    teams: list[TeamSummary]


class TeamStoreState(BaseModel):
    team_id: int
    team_abbr: str
    own: dict[str, Optional[float]]       # TEAM_* offensive context
    allowed: dict[str, Optional[float]]   # OPP_ALLOWED_* defensive (what they give up)


class InitRequest(BaseModel):
    season: Optional[str] = None


class PredictPlayerRequest(BaseModel):
    player_id: int
    minutes: float

File: frontend/src/types/simulation.ts
export interface StatCell {
  value: number;
  low?: number | null;
  high?: number | null;
}

export interface SimGame {
  game_id: string;
  team_id: number;
  team_abbr: string;
  opponent_team_id: number;
  opponent_abbr: string;
  matchup: string;
}

export interface SimState {
  season: string;
  current_date: string | null;
  next_game_day: string | null;
  day_index: number;
  total_days: number;
  finished: boolean;
  num_games: number;
  games: SimGame[];
}

export interface PlayerPrediction {
  player_id: number;
  player_name: string;
  team_id: number;
  team_abbr: string;
  opponent_team_id: number;
  opponent_abbr: string;
  is_home: boolean;
  minutes: number;
  default_minutes: number;
  eligible: boolean;
  status: 'green' | 'orange' | 'red';
  reason: string;
  stats: Record<string, StatCell>;
}

export interface EvalRow {
  player_id: number;
  player_name: string;
  team_abbr: string;
  opponent_abbr: string;
  is_home: boolean;
  real_minutes: number;
  eligible: boolean;
  reason: string;
  predicted: Record<string, number>;
  actual: Record<string, number>;
}

export interface LastResults {
  played_date: string | null;
  evaluations: EvalRow[];
}

export interface UpcomingResponse {
  state: SimState;
  predictions: PlayerPrediction[];
  last_results?: LastResults | null;
  resid_sigma?: Record<string, number>;
}

export interface AdvanceResponse {
  played_date: string | null;
  evaluations: EvalRow[];
  state: SimState;
}

export interface PlayerStoreSummary {
  player_id: number;
  player_name: string;
  team_abbr: string;
  games_count: number;
  eligible: boolean;
}

export interface PlayersListResponse {
  current_date: string | null;
  next_game_day: string | null;
  players: PlayerStoreSummary[];
}

export interface PlayerStoreState {
  player_id: number;
  player_name: string;
  team_abbr: string;
  position: string;
  last_game_date: string | null;
  games_count: number;
  eligible: boolean;
  features: Record<string, number | null>;
}

export interface TeamSummary {
  team_id: number;
  team_abbr: string;
}

export interface TeamsListResponse {
  teams: TeamSummary[];
}

export interface TeamStoreState {
  team_id: number;
  team_abbr: string;
  own: Record<string, number | null>;
  allowed: Record<string, number | null>;
}
